Aiken: commit validator

ch1bo opened the pull request:
Created with @KtorZ on the Hydra/Mithril workshop in Nantes.

:smiling_imp:  Write the vCommit validator in Aiken and produce a `plutus.json` blueprint

:smiling_imp: Add `Hydra.Plutus` module which loads validators from a `plutus.json` blueprint. 

:smiling_imp: Use the aiken created vCommit validator in `hydra-node`

:smiling_imp: Profit! Abort/Collect transactions do fit double the participants; but, due to execution memory limits, the maximum amount of parties got increased to 7.

TODO:

- [x] Use `plutus.json` for CI golden tests.
- [x] Make sure the `hydra-plutus` lib is recompiled whenever `plutus.json` changes.
  * This worked once today, but then not again
  
![image](https://github.com/input-output-hk/hydra/assets/2621189/9c0961e8-b1d2-4dab-b4c0-9e297e8cc3c7)
  
- [x] Remove the plutus-tx commit validator.
- [ ] Re-do the audit (or prove equivalence of this aiken script to the plutus-tx script)

---

<!-- Consider each and tick it off one way or the other -->
* [x] CHANGELOG updated or not needed
* [x] Documentation updated or not needed
* [x] Haddocks updated or not needed
* [x] No new TODOs introduced or explained herafter

ch1bo committed:
Add aiken to our nix develop shell

We intend to create one of the hydra-plutus scripts with aiken.

ch1bo committed:
Draft an aiken project in hydra-plutus

ch1bo committed:
Draft an empty commit aiken validator and commit plutus.json

ch1bo committed:
Create a module to load blueprint json

This separates the aiken build, which yields the plutus.json blueprint
and the haskell build which provides access to the compiled validators.

ch1bo committed:
Replace plutus-tx commit with the aiken commit script

Instead of using the plutus-tx "compiledScript", we use the
loaded-from-blueprint aiken script in the init validator and when
constructing/observing transactions in the hydra-node.

KtorZ committed:
implement 'ViaCollectCom' path in Aiken

KtorZ committed:
Implement 'ViaAbort' road on commit validator in Aiken.

ch1bo committed:
Try to make plutus.json a extra source file

This is not yet working and it does not correctly detect changes to plutus.json

ch1bo committed:
Decode commit datum as a record

Plutus-tx seems to be encoding tuples as a Constr on-chain.

ffakenz committed:
Recreate plutus scripts to make golden specs pass

Also format commit validator.

ffakenz committed:
Add golden spec to verify plutus blueprint

ffakenz committed:
Add commit errors to the validator script

ffakenz committed:
Add comments to the aiken commit validator script

ffakenz committed:
Remove commit validator

Also add TODO comments for things that are missing to be removed.

ffakenz committed:
Add test to check plutus blueprint hash against the code reference in script info

ffakenz commented on hydra-plutus/aiken.toml:
```diff
@@ -0,0 +1,9 @@
+name = "hydra/hydra-plutus"
+version = "0.0.0"
```
to review suggestions:
- [ ] 0.0.1
- [ ] 0.13.0
- [x] other

ffakenz replied:
Will be using 0.14.0

v0d1ch committed:
Bring back commit redeemer haskell type

- In order to compile and rebase against master we need this type
since currently there is no way to go from the openApi style of json
encoding for plutus redeemer to actual Plutus `Data` type.
CIP-57 is related to this conversion but this is ongoing work still.

v0d1ch committed:
Bump plutus dependencies after rebase

v0d1ch commented:
> Make sure the hydra-plutus lib is recompiled whenever plutus.json changes.

Note: Locally I always see recompilation when I change plutus.json file

v0d1ch commented:
This is related to work on this branch https://github.com/cardano-foundation/CIPs/tree/master/CIP-0057 
I am not sure if we want to keep (duplicated) commit datum and redeemer on the Haskell side until this CIP is implemented by the Plutus team. Problem we are having is that currently there is no way to go from `plutus.json` blueprint of commit datum/redeemer to equivalent in haskell where we need this to be able to build a commit tx.

v0d1ch committed:
Remove redundant imports

v0d1ch committed:
Alter plutus.json and aiken.lock

ffakenz committed:
Fix commit mutation tests on abort and collect-com

By regenerating the plutus.json file using -k flag during aiken build.

ffakenz committed:
Remove unit-test focus

ffakenz committed:
Replace the use of ByteArray for String on aiken commit validator script

This removes the need of using aiken/builtin function decode_utf8,
providing some small optimization.

Also recreated plutus scripts.

ffakenz committed:
Increase the number of parties for FanOut tx-cost

To correlate with the rest.

ffakenz committed:
Increase the maximum number of parties to 7

ffakenz committed:
Remove commit plutus validator as it is not needed anymore

Also removed its golden spec.

Finally added a note about a fragility being present while accessing
the commit validator script from the blueprintJSON.

ffakenz commented on hydra-plutus/src/Hydra/Contract/Commit.hs:
```diff
@@ -1,47 +1,23 @@
--- version 7 and 8, only plutus-core version 1.0.0 is available.
-{-# OPTIONS_GHC -fplugin-opt PlutusTx.Plugin:target-version=1.0.0 #-}
 
--- | The validator used to collect & open or abort a Head.
```
We are missing to update this module docs

ffakenz committed:
Remove not needed TODO

Because soon then Plutus team will provide such parsers.

ffakenz committed:
Upgrade aiken project version to match next release

ffakenz committed:
Update CHANGELOG

v0d1ch committed:
Add a CIP-57 todo

ghost commented:
Regarding this task: 
<img width="629" alt="Screenshot 2023-10-05 at 16 37 37" src="https://github.com/input-output-hk/hydra/assets/79840582/dac6977b-ef55-4757-89fe-2f170fad546b">

This is not something we can do ourselves quickly, I will check with auditors how long it might take and report here but this might block integration of this PR.

v0d1ch committed:
Remove a redundant import

v0d1ch committed:
Update changelog

ghost commented:
Not sure why the CI is red. I wonder if you should merge that PR straight away given the security implications: I have reached out to bcryptic for more information on how they could review aiken SCs. The commit script is simple enough but this will be an issue for the `Initial` or `Head` scripts which are much more involved.

v0d1ch committed:
Re-add golden tests for commit validator

abailly committed:
Dump log on aiken failure

abailly committed:
Dump logs on aiken process failure (take 2)

abailly committed:
Dump logs on aiken process failure (take 3)

v0d1ch committed:
Add aiken as a build dep to hydra-plutus tests

v0d1ch commented on CHANGELOG.md:
```diff
@@ -10,10 +10,17 @@ changes.
 
 ## [0.14.0] - UNRELEASED
 
+- Persist all sent messages as well as vector clock used internally by the
```
Why is this message here? This comes from the reliability network persistence so probably something was rebased wrong?

ffakenz replied:
wtf... removed!

v0d1ch commented on flake.nix:
```diff
@@ -31,7 +33,7 @@
         };
         hydraPackages = import ./nix/hydra/packages.nix {
-          inherit hydraProject system pkgs cardano-node;
+          inherit hydraProject system pkgs cardano-node aiken;
```
:heart:

v0d1ch commented on hydra-node/src/Hydra/Chain.hs:
```diff
@@ -48,7 +48,7 @@ maxMainnetLovelace = 100_000_000
 -- and on-chan validators (see 'computeCollectComCost' 'computeAbortCost')
 maximumNumberOfParties :: Int
-maximumNumberOfParties = 4
+maximumNumberOfParties = 7
```
:rocket:

ch1bo replied:
Should revisit this

v0d1ch commented on hydra-plutus/test/Hydra/Plutus/GoldenSpec.hs:
```diff
@@ -12,35 +12,85 @@
+        (_, _, _, aikenProcessHandle) <- createProcess aikenProcess
+        waitForProcess aikenProcessHandle
+
+      unless (result == ExitSuccess) $
```
We probably don't need this piece of code now that we fixed the `aiken` nix build input in the plutus tests.

ffakenz replied:
good catch! removed!

v0d1ch reviewed (changes requested):
The main thing is we have an entry in the changelog that doesn't belong in this PR. Other than that :ok_hand:

ffakenz committed:
Remove unnecessary check for aiken process in hydra-plutus GoldenSpecs

We don't need this piece of code now that we fixed the aiken nix build input in
the plutus tests.

Also removed wrong entry in CHANGELOG.

ghost commented:
As the aiken scripts are not audited, and while they are awaiting audit, one option would be to add a _feature flag_ to the hydra-node that would allow a user to decide, while we are still developing them, to use them or not. eg. something like `hydra-node run --use-aiken ...` that would select the Aiken scripts (when they are available) to post transactions. 
This is more work as we need to maintain the old `Commit` script, but it allows us to merge this PR and keep developing Aiken scripts without having to bother about diverging changes.

ghost commented:
We close the PR for now because we think the experiment is done and provided good results, but there's not pressing need to migrate everything to aiken

ch1bo commented on hydra-plutus/test/Hydra/Plutus/GoldenSpec.hs:
```diff
@@ -12,35 +12,84 @@
+        waitForProcess gitStatusProcessHandle
+      -- Read git log file and verify plutus.json did not change
+      gitLogContents <- decodeUtf8 <$> readFileBS gitLogFilePath
+      gitLogContents `shouldNotContain` "plutus.json"
```
Should: not run `git` (assumes a working copy), but rather run `aiken build` and assert freshly generated `plutus.json` contents vs the embedded one from `Hydra.Plutus`

ch1bo commented on hydra-plutus/scripts/vCommit.plutus:
Should: not need this file anymore

ch1bo commented on hydra-plutus/hydra-plutus.cabal:
```diff
@@ -97,13 +100,16 @@ library
+    , plutus-core          >=1.7 && <1.10
+    , plutus-ledger-api    >=1.7 && <1.10
+    , plutus-tx            >=1.7 && <1.10
+    , plutus-tx-plugin     >=1.7 && <1.10
```
Need to update these

ch1bo commented on flake.lock:
```diff
@@ -66,6 +66,26 @@
+        "narHash": "sha256-cOhHYQUt7TgtNkN+Am5eGWlPbKVL8Vsn7FDf1b1Psfo=",
+        "owner": "aiken-lang",
+        "repo": "aiken",
+        "rev": "78ff6cf178f9e7707d1a5297e4a774c9ab81d73e",
```
Should: use the latest aiken version
